lib/ABI/FunctionType/Layout.cpp

```diff
@@ -0,0 +1,963 @@
+    revng_assert(MaybeReturnValueSize != std::nullopt);
+    revng_assert(ReturnValue.Size == *MaybeReturnValueSize);
+
+    model::QualifiedType ReturnType = Function.ReturnType();
+    ReturnType.Qualifiers().emplace_back(PointerQualifier);
+
+    revng_assert(!ABI.GeneralPurposeReturnValueRegisters().empty());
+    auto FirstRegister = ABI.GeneralPurposeReturnValueRegisters()[0];
+    model::TypedRegister ReturnPointer(FirstRegister);
+    ReturnPointer.Type() = std::move(ReturnType);
+    NewType.ReturnValues().insert(std::move(ReturnPointer));
+  } else {
+    // The function returns `void`: no need to do anything special.
+  }
+
+  model::StructType StackArguments;
+  uint64_t CurrentStackOffset = 0;
+
+  // Now that return value is figured out, the arguments are next.
+  auto Arguments = distributeArguments(Function.Arguments(),
+                                       ReturnValue.SizeOnStack != 0);
+  std::size_t Index = 0;
+  for (const DistributedArgument &Distributed : Arguments) {
+    if (!Distributed.RepresentsPadding) {
+      // Transfer the register arguments first.
+      for (model::Register::Values Register : Distributed.Registers) {
+        model::NamedTypedRegister Argument(Register);
+
+        auto M = addNewTypes(*Binary);
+        bool SimpleArgument = Distributed.Registers.size() == 1
+                              && Distributed.SizeOnStack == 0;
+        if (SimpleArgument) {
+          const auto &ArgumentType = Function.Arguments().at(Index).Type();
+          revng_assert(*ArgumentType.size() <= ABI.getPointerSize());
+          Argument.Type() = chooseType(ArgumentType, Register, M);
+
+          // TODO: see what can be done to preserve names better
+          auto ArgumentName = Function.Arguments().at(Index).name();
+          if (llvm::StringRef{ ArgumentName.str() }.take_front(8) != "unnamed_")
+            Argument.CustomName() = ArgumentName;
+        } else {
+          // TODO: We can do a better job of preserving types here, for example,
+          //       if this argument is a struct, we can try to use field
+          //       types instead of just overriding everything with `Generic`.
+          Argument.Type() = { M.genericRegisterType(Register), {} };
+        }
+        M.commit();
+
+        NewType.Arguments().insert(std::move(Argument));
+      }
+
+      // Then stack arguments.
+      if (Distributed.SizeOnStack != 0) {
+        auto ArgumentIterator = Function.Arguments().find(Index);
+        revng_assert(ArgumentIterator != Function.Arguments().end());
+        const model::Argument &Argument = *ArgumentIterator;
+
+        // Round the next offset based on the natural alignment.
+        uint64_t Alignment = *ABI.alignment(Argument.Type());
+        if (CurrentStackOffset % Alignment != 0)
+          CurrentStackOffset += (Alignment - CurrentStackOffset % Alignment);
+        std::size_t InternalStackOffset = CurrentStackOffset;
```

Comment:
Are you sure this is the right place to compute this? Shouldn't this be like a property of `Distributed`?

Reply:
Yeah, considering it's computed at two places, it is indeed beneficial to move it up.

Reply:
I looked deeper into it. And I think I actually prefer this computation to stay here, at the outer-most layer of the encapsulation.

Let me elaborate. Currently, the `distributeArguments` looks something like (oversimplified):
```cpp
std::vector<Distributed> distributeArguments(const SortedVector<model::Argument> &ModelArguments) {
  std::vector<Distributed> Result;
  for (const auto &ModelArgument : ModelArguments)
    Result.emplace_back(distributeOne(/* describe the `ModelArgument` generically and pass it in */);
  return Result;
}
```
and I really prefer it to be super-simple, since the description it uses is not 100% trivial.

The downside of this is that the actual heavy lifter (`distributeOne`) knows nothing about any other arguments but the current one (which lets it be a lot simpler that it would have otherwise been and also be reused for other stuff, like return values).

As such, `Distributed` actually tells nothing about the actual position of any argument on the stack -- all it says is that one would need at least X bytes to represent it there, and then the caller (who's going to have to convert the `Distributed` to whichever type it uses internally, be it `abi::FunctionType::Layout::Argument` or `model::StructType`) takes care of of the specifics of aligning different arguments and inserting pre/post padding. This lets us keep the most complex (IMO) part of the converter slightly simpler by offloading some non-immediate complexity to the outside.

Downside is that this part then has to be duplicated, so what I propose is to just offload it into
```cpp
class abi::Definition {
// . . .
public:
  std::uint64_t alignedOffset(std::uint64_t Offset,
                              const model::QualifiedType &Type) {
    const std::uint64_t Alignment = *this->alignment(Type);
    if (Offset % Alignment != 0)
      return Offset + Alignment - Offset % Alignment;

    return Offset;
  }
};
```

What do you think?

Reply:
Works

Reply:
Done.